tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc

```diff
@@ -112,19 +112,33 @@ class AutoMixedPrecisionTest : public GrapplerTest {
     if (gpu_available_) {
       virtual_cluster_.reset(new SingleMachine(/* timeout_s = */ 10, 1, 1));
     } else {
-      DeviceProperties device_properties;
-      device_properties.set_type("GPU");
+      if( num_gpus > 0) {
+        DeviceProperties device_properties;
+        device_properties.set_type("GPU");
 #if GOOGLE_CUDA
-      device_properties.mutable_environment()->insert({"architecture", "7"});
-      device_properties.mutable_environment()->insert({"cuda", "9010"});
+        device_properties.mutable_environment()->insert({"architecture", "7"});
+        device_properties.mutable_environment()->insert({"cuda", "9010"});
 #else
-      device_properties.mutable_environment()->insert(
-          {"architecture", "gfx906"});
+        device_properties.mutable_environment()->insert(
+            {"architecture", "gfx906"});
 #endif
-      virtual_cluster_.reset(
-          new VirtualCluster({{"/GPU:1", device_properties}}));
+        virtual_cluster_.reset(
+            new VirtualCluster({{"/GPU:1", device_properties}}));
+      } else {
+	// When no GPUs are available, try running on CPU.
+        DeviceProperties device_properties;
+        device_properties.set_type("CPU");
+        virtual_cluster_.reset(new SingleMachine(/* timeout_s = */ 10, 1, 0));
+      }
     }
     TF_CHECK_OK(virtual_cluster_->Provision());
+
+    run_fp16_on_cpu_ = false;
```

Comment:
Since we only use this variable locally, it can just be a local variable in this function. (We can make it a class member later when there's a need.)

Reply:
done